File: godfather/roles/town/tracker.py
from godfather.roles.mixins import SingleAction, Townie
from godfather.game.types import Priority

DESCRIPTION = 'You may track one person at night to find out who they visited.'


class Tracker(Townie, SingleAction):
    name = 'Tracker'
    description = DESCRIPTION

    def __init__(self):
        super().__init__()
        self.action = 'track'
        self.action_gerund = 'tracking'
        self.action_priority = Priority.TRACKER
        self.action_text = 'track a player'

    async def tear_down(self, actions, player, target):
        visited_players = [
            visited.user.name for visited in actions.game.players if target in visited.visitors]
        if len(visited_players) >= 0:
            await player.user.send('Your target visited {}.'.format(', '.join(visited_players)))

File: src/lib/mafia/roles/town/Tracker.ts
import Townie from '@mafia/mixins/Townie';
import type Player from '@mafia/structures/Player';
import { TrackAction } from '../../actions/common/TrackAction';
import type { NightAction } from '../../managers/NightAction';
import { ActionRole } from '../../structures/ActionRole';

class Tracker extends ActionRole {
	public name = 'Tracker';
	public actions: NightAction[] = [new TrackAction(this)];

	public constructor(player: Player) {
		super(player);
		this.description = this.game.t('roles/town:trackerDescription');
	}
}

Tracker.categories = [...Tracker.categories, 'Town Investigative'];

export default Townie(Tracker);
